ostd/src/mm/kspace/kva.rs

```diff
@@ -0,0 +1,456 @@
+    fn free(&mut self, range: Range<Vaddr>) {
+        // 1. get the previous free block, check if we can merge this block with the free one
+        //     - if contiguous, merge this area with the free block.
+        //     - if not contiguous, create a new free block, insert it into the list.
+        let mut free_range = range.clone();
+
+        if let Some((prev_va, prev_node)) = self
+            .freelist
+            .upper_bound_mut(core::ops::Bound::Excluded(&free_range.start))
+            .peek_prev()
+        {
+            if prev_node.block.end == free_range.start {
+                let prev_va = *prev_va;
+                free_range.start = prev_node.block.start;
+                self.freelist.remove(&prev_va);
+            }
+        }
+        self.freelist
+            .insert(free_range.start, KvaFreeNode::new(free_range.clone()));
+
+        // 2. check if we can merge the current block with the next block, if we can, do so.
+        if let Some((next_va, next_node)) = self
+            .freelist
+            .lower_bound_mut(core::ops::Bound::Excluded(&free_range.start))
+            .peek_next()
+        {
+            if free_range.end == next_node.block.start {
+                let next_va = *next_va;
+                free_range.end = next_node.block.end;
+                self.freelist.remove(&next_va);
+                self.freelist.get_mut(&free_range.start).unwrap().block.end = free_range.end;
+            }
+        }
+    }
+}
+
+lazy_static! {
+    pub static ref KVA_TRACKED_ALLOCATOR: SpinLock<VirtAddrAllocator> =
+        SpinLock::new(VirtAddrAllocator::new(TRACKED_MAPPED_PAGES_RANGE));
+    pub static ref KVA_UNTRACKED_ALLOCATOR: SpinLock<VirtAddrAllocator> =
+        SpinLock::new(VirtAddrAllocator::new(VMALLOC_VADDR_RANGE));
+}
+
+#[derive(Debug)]
+pub enum KvaType {
+    Tracked,
+    UnTracked,
+}
+
+#[derive(Debug)]
+pub struct Kva {
+    range: Range<Vaddr>,
+    num_cpus_: u32,
+    type_: KvaType,
+}
+
+impl Kva {
+    pub fn new(size: usize, type_: KvaType) -> Self {
+        let mut lock_guard = match type_ {
+            KvaType::Tracked => KVA_TRACKED_ALLOCATOR.lock(),
+            KvaType::UnTracked => KVA_UNTRACKED_ALLOCATOR.lock(),
+        };
+        let range = lock_guard.deref_mut().alloc(size).unwrap();
+        let num_cpus_ = num_cpus();
+        Kva {
+            range,
+            num_cpus_,
+            type_,
+        }
+    }
+
+    pub fn start(&self) -> Vaddr {
+        self.range.start
+    }
+
+    pub fn end(&self) -> Vaddr {
+        self.range.end
+    }
+
+    pub fn range(&self) -> Range<Vaddr> {
+        self.range.start..self.range.end
+    }
+
+    /// Map pages into the kernel virtual area.
+    /// # Safety
+    /// The caller should ensure either the mapped pages or the range to be used doesn't
+    /// violate the memory safety of kernel objects.
+    pub unsafe fn map_pages<T: PageMeta>(&mut self, range: Range<Vaddr>, pages: Vec<Page<T>>) {
```

Comment:
You can use `impl Iterator<Item = Page<T>>` instead of using `Vec<Page<T>>`. Then the user is free to not use Vec. Vec introduces some unwanted heap allocations. #1287 implements iterator for `Segment` so there's chance to eliminate heap allocations for it.